Homework 8

mrnchernova opened the pull request.

Commit:
homework 8: Search note in notebook

Commit:
homework 8: TimeInterval ---

Commit:
homework 8: finished task for searching time interval

Commit:
homework 8: ATM.java --

Commit:
homework 8: ATM.java --

Commit:
homework 8: ATM task solved.
PS don't beat me for this solution

DimaLamaka commented on part_1/src/com/chernova/homework_8/TimeInterval.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    // метод для получения полного количества секунд в объекте
+    public static int secondsOnly(Time t) {
```
наименование метода должно означать действие
и метод сам для чего? либо он конвертирует все время в объекте в секунды, либо он у объекта берет все время в секундах и возвращает ?

DimaLamaka commented on part_1/src/com/chernova/homework_8/TimeInterval.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    // перевод каких бы то ни было введенных данных в нормальное представление часов, минут, секунд
+    public static Time correctDate(Time t) {
```
исправь наименование аргумента, на time хотя бы, для читабельности

DimaLamaka commented on part_1/src/com/chernova/homework_8/TimeInterval.java:
```diff
@@ -0,0 +1,87 @@
+        ArrayList<Time> t = new ArrayList<>();
+        ArrayList<Integer> sec = new ArrayList<>();
+
+        Time a = new Time(120);
```
наименование переменных

DimaLamaka commented on part_1/src/com/chernova/homework_8/TimeInterval.java:
```diff
@@ -0,0 +1,87 @@
+
+public class TimeInterval {
+    public static void main(String[] args) {
+        ArrayList<Time> t = new ArrayList<>();
```
Для чего эти листы?

mrnchernova replied:
Были какие-то планы на них. Забыла удалить

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+
+    public static void printNote(ArrayList<Note> notes, String s) {
```
наименование аргумента String s

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -0,0 +1,76 @@
+    public static void printNote(ArrayList<Note> notes, String s) {
+        boolean isMatch = false;
+
+        if (s.equals("")) {
```
если s будет null, то приложение упадет с эксепшеном

mrnchernova replied:
Да, но в этом решении s никогда не будет null. В начале программы  s проинициализирована

DimaLamaka replied:
файкт в том, что лучше у "" вызывать equals, что бы обезапасить себя, плюс наимование s поправь, на более читабельное

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -0,0 +1,76 @@
+        if (s.equals("")) {
+            for (int i = 0; i < notes.size(); i++) {
+                System.out.println(notes.get(i));
+            }
```
Можешь глянуть стаический метод Arrays.toString() , может будет удобнее:)

mrnchernova replied:
А разве этот метод не для массивов?

DimaLamaka replied:
да для массивов, тогда в твоем случае еще проще) можно просто в вывод передать сам лист, он вызовет toString() листа

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -0,0 +1,76 @@
+        } else {
+            for (int i = 0; i < notes.size(); i++) {
+                Pattern p = Pattern.compile(s.trim().toLowerCase(Locale.ROOT));
+                Matcher m = p.matcher(notes.get(i).note.toLowerCase(Locale.ROOT));
```
а не проще использовать метод contains() у класса String ?

mrnchernova replied:
Проще) Переделано

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -0,0 +1,76 @@
+        printNote(notes, s);
+
+        System.out.println("\nВведите слово для поиска:");
+        s = sc.nextLine();
```
сканер нужно закрывать

DimaLamaka commented on part_1/src/com/chernova/homework_8/ATM.java:
```diff
@@ -0,0 +1,192 @@
+        int total = byn.cash100 * 100 + byn.cash50 * 50 + byn.cash20 * 20;
+        System.out.println("Deposit money: " + total + " BYN");
+
+        int m = 0;  // сумма для снятия или пополнения
```
Наименование

DimaLamaka commented on part_1/src/com/chernova/homework_8/ATM.java:
```diff
@@ -0,0 +1,192 @@
+
+        System.out.println("Do you want to withdraw from an ATM or deposit cash in an ATM?");
+        System.out.println("w/d ?");
+        String s = sc.nextLine();
```
Наименование

DimaLamaka commented on part_1/src/com/chernova/homework_8/ATM.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+
+    public static void deposit(int m, int total, boolean solution) {
```
тут должна быть работа с банкоматом, то-есть самим объектом Money

DimaLamaka replied:
и если передавать отрицательное число ? я буду снимать деньги уже?)

DimaLamaka commented on part_1/src/com/chernova/homework_8/ATM.java:
```diff
@@ -0,0 +1,192 @@
+        int cash100 = byn.cash100;
+        int cash50 = byn.cash50;
+        int cash20 = byn.cash20;
+        int mCopy = m; // так как 'm' в процессе будет изменятся, то делаем ее копию
```
снятие денег должно производиться с полями объекта Money, а то сейчас я захочу снять денег, деньги сниму, а на счёте останется та же сумма, что только для меня хорошо:)

DimaLamaka commented on part_1/src/com/chernova/homework_8/ATM.java:
```diff
@@ -0,0 +1,192 @@
+            System.out.println(solution);
+            total += m;
+            System.out.println("Deposit money: " + total + " BYN");
+        }
```
у тебя номиналы в банкомате 100, 50, 20 ты не можешь брать другие

DimaLamaka commented on part_1/src/com/chernova/homework_8/ATM.java:
```diff
@@ -0,0 +1,192 @@
+                cash20--;
+                currency.append(20 + " ");
+            }
+
```
если я захочу снять 160, мне вернется ошибка, хотя я могу получить 1 - 100, 3 - 20

mrnchernova replied:
исправлено

Commit:
homework 8: fixed issues

DimaLamaka commented on part_1/src/com/chernova/homework_8/TimeInterval.java:
```diff
@@ -13,45 +11,43 @@
-        Time a = new Time(120);
-        Time b = new Time(1, 30, 5);
+        Time time_1 = new Time(120);
+        Time time_2 = new Time(1, 30, 5);
```
никаких нижних подчеркиваний в наименовании, только в константах они допустимы

DimaLamaka commented on part_1/src/com/chernova/homework_8/TimeInterval.java:
```diff
@@ -0,0 +1,83 @@
+        }
+
+        @Override
+        public int compareTo(Time time) {
```
для сравнения, лучше добавь поле в котором будет храниться общее кол-во секунд, и по нему сравнивай два объекта, или сравнивай по всем полям

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -0,0 +1,78 @@
+
+                if (notes.get(i).note.contains(search)) {
+                    System.out.println(notes.get(i));
+                    isMatch = true;
```
переменная isMatch лишняя, если нашли элемент, ты можешь распечатать его и сделать return;

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -0,0 +1,78 @@
+
+        if (!isMatch || search.equals("")) {
+            System.out.println("Совпадений не найдено");
+        }
```
а тут тогда не будет проверок и просто будет печать

Commit:
homework 8: ATM.java исправлено:
-добавление денег в банкомат;
-ввод отрицательных значений.
NotebookSearch.java:
-исправоен метод searchNote
TimeInterval.java:
-в объект Time добавлено поле, в которое заносится общее количество секунд

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -25,40 +25,42 @@ public static void main(String[] args) {
-            System.out.println(notes.get(i));
+            list.add(notes.get(i));
         }
+        return list;
```
странный метод, по названию он для распечатки листа, а в логике его, просто ты перегоняешь элементы с одного листа в другой лист  , зачем ?)

mrnchernova replied:
Раньше он распечатывал список задач чтоб красиво было. А теперь я научилась делать это в одну строку.
Удаляю

DimaLamaka commented on part_1/src/com/chernova/homework_8/NotebookSearch.java:
```diff
@@ -25,40 +25,42 @@ public static void main(String[] args) {
+    public static ArrayList<Note> searchNote(ArrayList<Note> notes, String search) {
+        ArrayList<Note> list = new ArrayList<>();
 
         if (!search.equals("")) {
```
у класса String, есть метод isEmpty(), а если ты и сравниваешь строки, то лучше вызывай equals в строке, где она точно будет, потому что если search придет null то у тебя ошибка будет в этой строке

DimaLamaka commented on part_1/src/com/chernova/homework_8/TimeInterval.java:
```diff
@@ -43,18 +43,16 @@ public static Time correctDate(Time time) {
-        return time.seconds;
+    public static int allTimeInSeconds(Time time) {
+        return time.allTimeInSeconds = (time.seconds + time.minutes * 60 + time.hours * 3600);
     }
```
а зачем тогда возвращаемый тип, если ты в поле time, просто задаешь значение

mrnchernova replied:
что-то я не догоняю в чем ошибка:(

DimaLamaka commented on part_1/src/com/chernova/homework_8/ATM.java:
```diff
@@ -38,7 +40,6 @@ public static void main(String[] args) {
 
         switch (s) {
```
старайся давать более читабельные наименования

Commit:
homework 8: removed method printNote() in NotebookSearch.java;
fixed some issues in ATM.java
